matrix_sdk/src/event_handler.rs

```diff
@@ -0,0 +1,397 @@
+    fn from_data(data: &EventHandlerData<'_>) -> Option<Self> {
+        data.room.clone()
+    }
+}
+
+/// The raw JSON form of an event.
+///
+/// Used as a context argument for event handlers (see
+/// [`Client::register_event_handler`]).
+// FIXME: This could be made to not own the raw JSON value with some changes to
+//        the traits above, but only with GATs.
+#[derive(Clone, Debug)]
+pub struct RawEvent(pub Box<RawJsonValue>);
+
+impl Deref for RawEvent {
+    type Target = RawJsonValue;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl EventHandlerContext for RawEvent {
+    fn from_data(data: &EventHandlerData<'_>) -> Option<Self> {
+        Some(Self(data.raw.to_owned()))
+    }
+}
+
+/// Return types supported for event handlers implement this trait.
+///
+/// It is not meant to be implemented outside of matrix-sdk.
+pub trait EventHandlerResult: Sized {
+    #[doc(hidden)]
+    fn print_error(&self, event_type: &str);
+}
+
+impl EventHandlerResult for () {
+    fn print_error(&self, _event_type: &str) {}
+}
+
+impl<E: std::error::Error> EventHandlerResult for Result<(), E> {
+    fn print_error(&self, event_type: &str) {
+        if let Err(e) = self {
+            tracing::error!("Event handler for `{}` failed: {}", event_type, e);
+        }
+    }
+}
+
+#[cfg(feature = "anyhow")]
+impl EventHandlerResult for anyhow::Result<()> {
+    fn print_error(&self, event_type: &str) {
+        if let Err(e) = self {
+            tracing::error!("Event handler for `{}` failed: {:?}", event_type, e);
+        }
+    }
+}
+
+#[derive(Deserialize)]
+struct UnsignedDetails {
+    redacted_because: Option<serde::de::IgnoredAny>,
+}
+
+impl Client {
+    pub(crate) async fn handle_sync_events<T>(
+        &self,
+        kind: EventKind,
+        room: &Option<room::Room>,
+        events: &[Raw<T>],
+    ) -> serde_json::Result<()> {
+        self.handle_sync_events_wrapped(kind, room, events, |x| x).await
+    }
+
+    pub(crate) async fn handle_sync_state_events(
+        &self,
+        room: &Option<room::Room>,
+        state_events: &[Raw<AnySyncStateEvent>],
+    ) -> serde_json::Result<()> {
+        #[derive(Deserialize)]
+        struct StateEventDetails<'a> {
+            #[serde(borrow, rename = "type")]
+            event_type: Cow<'a, str>,
+            unsigned: Option<UnsignedDetails>,
+        }
+
+        self.handle_sync_events_wrapped_with(room, state_events, std::convert::identity, |raw| {
+            let StateEventDetails { event_type, unsigned } = raw.deserialize_as()?;
+            let redacted = unsigned.and_then(|u| u.redacted_because).is_some();
+            Ok((EventKind::State { redacted }, event_type))
+        })
+        .await
+    }
+
+    pub(crate) async fn handle_sync_timeline_events(
+        &self,
+        room: &Option<room::Room>,
+        timeline_events: &[SyncRoomEvent],
+    ) -> serde_json::Result<()> {
+        // FIXME: add EncryptionInfo to context
```

Comment:
Do you plan on tackling that as part of the PR?

Reply:
If you prefer yes, otherwise I'll probably create a followup PR soon-ish.

Reply:
A followup PR is fine.